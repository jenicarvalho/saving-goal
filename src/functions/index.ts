export const formatCurrency = (value: number): string => {
  return (value / 100).toLocaleString('en-US', {
    style: 'currency',
    currency: 'USD',
  });
};
