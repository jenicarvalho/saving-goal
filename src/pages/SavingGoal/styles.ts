import styled from 'styled-components';

export const goal = styled.div`
  display: flex;
  align-items: center;

  img {
    margin-right: 16px;
  }

  h2 {
    font-size: 24px;
    color: var(--blueGray900);
    font-family: Rubik, sans-serif;
    position: relative;
    bottom: 4px;
  }

  span {
    color: var(--blueGray400);
    font-size: 16px;
  }

  @media (max-width: 414px) {
    h2 {
      font-size: 20px;
    }

    span {
      font-size: 14px;
    }
  }
`;

export const inputs = styled.div`
  display: flex;
  margin-top: 30px;

  label {
    font-size: 14px;
    color: var(--blueGray900);
    margin-bottom: 4px;
  }

  > div:nth-child(2) {
    margin-left: 16px;
  }

  @media (max-width: 414px) {
    display: block;

    > div:nth-child(2) {
      margin: 16px 0 0 0;
    }
  }
`;

export const singleInput = styled.div`
  display: flex;
  flex-direction: column;
`;

export const monthlyAmount = styled.div`
  margin-top: 24px;
  border-radius: 8px;
  border: 1px solid #e9eef2;
  color: var(--blueGray900);

  h3 {
    font-size: 20px;
    font-weight: normal;
  }

  span {
    font-family: Rubik, sans-serif;
    color: var(--brandColorSecondary);
    font-size: 32px;
    font-weight: bold;
  }
`;

export const amount = styled.div`
  padding: 20px 30px;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const result = styled.div`
  padding: 20px 30px;
  font-size: 12px;
  line-height: 16px;
  background-color: var(--blueGray10);
`;
