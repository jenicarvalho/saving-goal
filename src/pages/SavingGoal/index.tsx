import { useState, useCallback } from 'react';
import addMonths from 'date-fns/addMonths';
import differenceInMonths from 'date-fns/differenceInMonths';

import { formatCurrency } from '../../functions';
import * as S from './styles';
import {
  container as Container,
  button as Button,
  H1,
} from '../../styles/global';
import Card from '../../components/Card';
import house from '../../assets/icons/buy-a-house.svg';
import Money from '../../components/Inputs/Money';
import DateField from '../../components/Inputs/DateField';

function SavingGoal(): JSX.Element {
  const [amount, setAmount] = useState(0);
  const [reachDate, setReachDate] = useState<Date>(addMonths(new Date(), 1));
  const handleAmountChange = useCallback((amount): void => {
    setAmount(amount);
  }, []);

  const handleReachDateChange = useCallback((reachDate): void => {
    setReachDate(reachDate);
  }, []);

  const months = differenceInMonths(reachDate, new Date());
  const saving = Number(amount) / Number(months + 1);
  const savingFormated = formatCurrency(saving);

  return (
    <Container>
      <H1>
        Lets plan your <strong>saving goal</strong>
      </H1>

      <Card>
        <S.goal>
          <img src={house} alt="House goal" />
          <div>
            <h2>Buy a house</h2>
            <span>Saving goal</span>
          </div>
        </S.goal>

        <S.inputs>
          <S.singleInput>
            <label>Total amount</label>
            <Money onAmountChange={handleAmountChange} value={amount} />
          </S.singleInput>
          <S.singleInput>
            <label>Reach goal by</label>
            <DateField
              onReachDateChange={handleReachDateChange}
              value={reachDate}
            />
          </S.singleInput>
        </S.inputs>

        <S.monthlyAmount>
          <S.amount>
            <h3>Monthly amount</h3>
            <span id="result-formated">{savingFormated}</span>
          </S.amount>
          <S.result>
            Youâ€™re planning <strong>{months + 1} monthly deposits</strong> to
            reach your <strong>{formatCurrency(amount)}</strong> goal by{' '}
            <strong>
              {reachDate.toLocaleString('default', {
                month: 'long',
                year: 'numeric',
              })}
            </strong>
            .
          </S.result>
        </S.monthlyAmount>

        <Button>
          <button>Confirm</button>
        </Button>
      </Card>
    </Container>
  );
}

export default SavingGoal;
