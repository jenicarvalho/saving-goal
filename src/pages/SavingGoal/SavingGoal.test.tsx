import SavingGoal from '.';
import { fireEvent, render } from '@testing-library/react';
import addMonths from 'date-fns/addMonths';
import differenceInMonths from 'date-fns/differenceInMonths';
import { formatCurrency } from '../../functions';

describe('<SavingGoal />', () => {
  it('render the page', () => {
    const component = render(<SavingGoal />);
    expect(component).toMatchSnapshot();
  });
  it('show the correct monthly amount', () => {
    const component = render(<SavingGoal />);

    const { container } = component;

    // money component
    const amount = 50000;
    const inputAmount = component.getByTestId('input-amount');

    fireEvent.keyDown(inputAmount, { key: 5, code: 53, charCode: 53 });
    fireEvent.keyDown(inputAmount, { key: 0, code: 48, charCode: 48 });
    fireEvent.keyDown(inputAmount, { key: 0, code: 48, charCode: 48 });
    fireEvent.keyDown(inputAmount, { key: 0, code: 48, charCode: 48 });
    fireEvent.keyDown(inputAmount, { key: 0, code: 48, charCode: 48 });

    // reachDate component
    const reachDate = addMonths(new Date(), 4);
    const inputReachDate = container.querySelector('input[name="reachDate"]');
    fireEvent.change(inputReachDate, { target: { value: reachDate } });

    const months = differenceInMonths(reachDate, new Date());
    const saving = Number(amount) / Number(months + 1);
    const savingFormated = formatCurrency(saving);
    const result = container.querySelector('#result-formated');

    expect(result).toHaveTextContent(savingFormated);
  });
});
