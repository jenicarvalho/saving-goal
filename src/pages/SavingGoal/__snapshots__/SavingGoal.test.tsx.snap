// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SavingGoal /> render the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gKAaRy jzCGqt"
      >
        <h1
          class="sc-iCoGMd kreHou"
        >
          Lets plan your 
          <strong>
            saving goal
          </strong>
        </h1>
        <div
          class="sc-pNWdM fQQXVb"
        >
          <div
            class="sc-bdnxRM chcKIG"
          >
            <img
              alt="House goal"
              src="buy-a-house.svg"
            />
            <div>
              <h2>
                Buy a house
              </h2>
              <span>
                Saving goal
              </span>
            </div>
          </div>
          <div
            class="sc-gtsrHT gkqZGB"
          >
            <div
              class="sc-dlnjwi koSwQA"
            >
              <label>
                Total amount
              </label>
              <div
                class="sc-jrsJWt liznJd"
              >
                <svg
                  fill="none"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12 0C12.5523 0 13 0.447715 13 1V23C13 23.5523 12.5523 24 12 24C11.4477 24 11 23.5523 11 23V1C11 0.447715 11.4477 0 12 0Z"
                    fill="#CBD5DC"
                    fill-rule="evenodd"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M6.31802 5.31802C7.16193 4.47411 8.30653 4 9.5 4H17C17.5523 4 18 4.44772 18 5C18 5.55228 17.5523 6 17 6H9.5C8.83696 6 8.20107 6.26339 7.73223 6.73223C7.26339 7.20107 7 7.83696 7 8.5C7 9.16304 7.26339 9.79893 7.73223 10.2678C8.20107 10.7366 8.83696 11 9.5 11H14.5C15.6935 11 16.8381 11.4741 17.682 12.318C18.5259 13.1619 19 14.3065 19 15.5C19 16.6935 18.5259 17.8381 17.682 18.682C16.8381 19.5259 15.6935 20 14.5 20H6C5.44772 20 5 19.5523 5 19C5 18.4477 5.44772 18 6 18H14.5C15.163 18 15.7989 17.7366 16.2678 17.2678C16.7366 16.7989 17 16.163 17 15.5C17 14.837 16.7366 14.2011 16.2678 13.7322C15.7989 13.2634 15.163 13 14.5 13H9.5C8.30653 13 7.16193 12.5259 6.31802 11.682C5.47411 10.8381 5 9.69347 5 8.5C5 7.30653 5.47411 6.16193 6.31802 5.31802Z"
                    fill="#CBD5DC"
                    fill-rule="evenodd"
                  />
                </svg>
                <input
                  data-testid="input-amount"
                  name="amount"
                  type="text"
                  value="0.00"
                />
              </div>
            </div>
            <div
              class="sc-dlnjwi koSwQA"
            >
              <label>
                Reach goal by
              </label>
              <div
                class="sc-jrsJWt liznJd"
              >
                <div
                  class="react-datepicker-wrapper"
                >
                  <div
                    class="react-datepicker__input-container"
                  >
                    <input
                      class=""
                      name="reachDate"
                      type="text"
                      value="February"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-hKFxyN uJSbO"
          >
            <div
              class="sc-eCApnc fPOqUF"
            >
              <h3>
                Monthly amount
              </h3>
              <span
                id="result-formated"
              >
                $0.00
              </span>
            </div>
            <div
              class="sc-jSFjdj igEdBi"
            >
              You’re planning 
              <strong>
                2
                 monthly deposits
              </strong>
               to reach your 
              <strong>
                $0.00
              </strong>
               goal by
               
              <strong>
                February 2022
              </strong>
              .
            </div>
          </div>
          <div
            class="sc-fujyAs dgAbqu"
          >
            <button>
              Confirm
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gKAaRy jzCGqt"
    >
      <h1
        class="sc-iCoGMd kreHou"
      >
        Lets plan your 
        <strong>
          saving goal
        </strong>
      </h1>
      <div
        class="sc-pNWdM fQQXVb"
      >
        <div
          class="sc-bdnxRM chcKIG"
        >
          <img
            alt="House goal"
            src="buy-a-house.svg"
          />
          <div>
            <h2>
              Buy a house
            </h2>
            <span>
              Saving goal
            </span>
          </div>
        </div>
        <div
          class="sc-gtsrHT gkqZGB"
        >
          <div
            class="sc-dlnjwi koSwQA"
          >
            <label>
              Total amount
            </label>
            <div
              class="sc-jrsJWt liznJd"
            >
              <svg
                fill="none"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12 0C12.5523 0 13 0.447715 13 1V23C13 23.5523 12.5523 24 12 24C11.4477 24 11 23.5523 11 23V1C11 0.447715 11.4477 0 12 0Z"
                  fill="#CBD5DC"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M6.31802 5.31802C7.16193 4.47411 8.30653 4 9.5 4H17C17.5523 4 18 4.44772 18 5C18 5.55228 17.5523 6 17 6H9.5C8.83696 6 8.20107 6.26339 7.73223 6.73223C7.26339 7.20107 7 7.83696 7 8.5C7 9.16304 7.26339 9.79893 7.73223 10.2678C8.20107 10.7366 8.83696 11 9.5 11H14.5C15.6935 11 16.8381 11.4741 17.682 12.318C18.5259 13.1619 19 14.3065 19 15.5C19 16.6935 18.5259 17.8381 17.682 18.682C16.8381 19.5259 15.6935 20 14.5 20H6C5.44772 20 5 19.5523 5 19C5 18.4477 5.44772 18 6 18H14.5C15.163 18 15.7989 17.7366 16.2678 17.2678C16.7366 16.7989 17 16.163 17 15.5C17 14.837 16.7366 14.2011 16.2678 13.7322C15.7989 13.2634 15.163 13 14.5 13H9.5C8.30653 13 7.16193 12.5259 6.31802 11.682C5.47411 10.8381 5 9.69347 5 8.5C5 7.30653 5.47411 6.16193 6.31802 5.31802Z"
                  fill="#CBD5DC"
                  fill-rule="evenodd"
                />
              </svg>
              <input
                data-testid="input-amount"
                name="amount"
                type="text"
                value="0.00"
              />
            </div>
          </div>
          <div
            class="sc-dlnjwi koSwQA"
          >
            <label>
              Reach goal by
            </label>
            <div
              class="sc-jrsJWt liznJd"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class=""
                    name="reachDate"
                    type="text"
                    value="February"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-hKFxyN uJSbO"
        >
          <div
            class="sc-eCApnc fPOqUF"
          >
            <h3>
              Monthly amount
            </h3>
            <span
              id="result-formated"
            >
              $0.00
            </span>
          </div>
          <div
            class="sc-jSFjdj igEdBi"
          >
            You’re planning 
            <strong>
              2
               monthly deposits
            </strong>
             to reach your 
            <strong>
              $0.00
            </strong>
             goal by
             
            <strong>
              February 2022
            </strong>
            .
          </div>
        </div>
        <div
          class="sc-fujyAs dgAbqu"
        >
          <button>
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
